### PR TITLE
issue: 1450183 Remove Ring duplications

### PR DESCRIPTION
**Pull request by liranoz12**

### CODE REVIEW

File: src/vma/dev/ring_slave.cpp [hunk 1]
```diff
@@ -222,6 +224,22 @@ void ring_slave::flow_tcp_del_all()
...
+
+	ring_logfuncall("Allocating additional %d buffers for internal use", count);
+
+	bool res = global_pool->get_buffers_thread_safe(local_pool,
+			this, count, lkey);
+	if (!res) {
+		ring_logfunc("Out of mem_buf_desc from %s free pool for internal object pool", is_rx ? "rx" : "tx");
+		return false;
+	}
+
```
Comment (igor-ivanov) [inline src/vma/dev/ring_slave.cpp]: Is statistic missed?
https://github.com/Mellanox/libvma/pull/695/commits/ee02fea91aaaddee4a43a91c6bd80e2318c6e7a9#diff-1e15a9c45d43c8e00766447137365b1fL335

File: src/vma/dev/ring_slave.cpp [hunk 2]
```diff
@@ -222,6 +224,22 @@ void ring_slave::flow_tcp_del_all()
 	}
 }
 
+bool ring_slave::request_more_buffers(descq_t &local_pool, uint32_t count, uint32_t lkey, bool is_rx)
```
Comment (igor-ivanov) [inline src/vma/dev/ring_slave.cpp]: Please consider two inline functions as `request_more_rx_buffers` and `request_more_tx_buffers`
It is improve code readability and exclude if condition

File: src/vma/dev/ring.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@
 #define MODULE_HDR      MODULE_NAME "%d:%s() "
 
 ring::ring() :
-	m_p_n_rx_channel_fds(NULL), m_parent(NULL), m_is_mp_ring(false)
+	m_p_n_rx_channel_fds(NULL), m_parent(NULL)
```
Comment (igor-ivanov) [inline src/vma/dev/ring.cpp]: please start using single style as "one member initialization per line"

File: src/vma/dev/ring_slave.h [hunk 4]
```diff
@@ -162,59 +162,55 @@ class ring_slave : public ring
...
-	virtual ring_user_id_t	generate_id(const address_t src_mac, const address_t dst_mac, uint16_t eth_proto, uint16_t encap_proto, uint32_t src_ip, uint32_t dst_ip, uint16_t src_port, uint16_t dst_port);
-	virtual bool        is_up() = 0;
-	virtual void        inc_tx_retransmissions(ring_user_id_t id);
-	virtual void        inc_cq_moderation(size_t sz_data) = 0;
-	virtual bool        rx_process_buffer(mem_buf_desc_t* p_rx_wc_buf_desc, void* pv_fd_ready_array);
-	virtual uint32_t    get_underly_qpn() = 0;
 	virtual bool        attach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink);
 	virtual bool        detach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink);
+	virtual bool        rx_process_buffer(mem_buf_desc_t* p_rx_wc_buf_desc, void* pv_fd_ready_array);
+	virtual void        print_val();
```
Comment (igor-ivanov) [inline src/vma/dev/ring_slave.h]: I prefer to see print_val() after constructor and destructor.

File: src/vma/dev/ring.cpp (hunk 3 above)
Comment (liranoz12) [reply src/vma/dev/ring.cpp]: done

File: src/vma/dev/ring_slave.h (hunk 4 above)
Comment (liranoz12) [reply src/vma/dev/ring_slave.h]: done

File: src/vma/dev/ring_slave.cpp (hunk 1 above)
Comment (liranoz12) [reply src/vma/dev/ring_slave.cpp]: no, it is updated from compensate_qp_poll_success. from example :
https://github.com/Mellanox/libvma/commit/ee02fea91aaaddee4a43a91c6bd80e2318c6e7a9#diff-7d5519a685a26b0a2b28191d5eac7783L64

File: src/vma/dev/ring_slave.cpp [hunk 5]
```diff
@@ -517,3 +536,442 @@ bool ring_slave::detach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink)
 
 	return true;
 }
+
+// calling sockinfo callback with RFS bypass
+static inline bool check_rx_packet(sockinfo *si, mem_buf_desc_t* p_rx_wc_buf_desc, void *fd_ready_array)
+{
+	// Dispatching: Notify new packet to the FIRST registered receiver ONLY
+#ifdef RDTSC_MEASURE_RX_DISPATCH_PACKET
```
[the remark this answers is not in the document]
Comment (liranoz12) [reply src/vma/dev/ring_slave.cpp]: it is broken because g_rdtsc_instr_info_arr is defined only for socketXtreme

**Comment by liranoz12** [comment]
bot:retest

**Comment by liranoz12** [comment]
bot:retest

**Comment by liranoz12** [comment]
bot:retest

File: src/vma/dev/ring_slave.cpp (hunk 2 above)
Comment (liranoz12) [reply src/vma/dev/ring_slave.cpp]: i reverted this commit

**Pull request merged**